okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/HappyEyeballsExchangeFinder.kt

```diff
@@ -0,0 +1,135 @@
+import okhttp3.internal.concurrent.TaskRunner
+import okhttp3.internal.connection.RoutePlanner.Plan
+import okhttp3.internal.okHttpName
+
+/**
+ * Speculatively connects to each IP address of a target address, returning as soon as one of them
+ * connects successfully. This kicks off new attempts every 250 ms until a connect succeeds.
+ */
+internal class HappyEyeballsExchangeFinder(
+  private val routePlanner: RoutePlanner,
+  private val taskRunner: TaskRunner,
+) {
+  private val connectDelayMillis = 250L
+
+  /** Plans currently being connected, and that will later be added to [connectResults]. */
+  private var connectsInFlight = mutableListOf<Plan>()
+
+  /**
+   * Results are posted here as they occur. The find job is done when either one plan completes
+   * successfully or all plans fail.
+   */
+  private val connectResults = LinkedBlockingDeque<ConnectResult>()
+
+  /** Exceptions accumulate here. */
+  private var firstException: IOException? = null
+
+  /** True until we've launched all the connects we'll ever launch. */
+  private var morePlansExist = true
+
+  fun find(): RealConnection {
+    try {
+      while (morePlansExist || connectsInFlight.isNotEmpty()) {
+        if (routePlanner.isCanceled()) throw IOException("Canceled")
+
+        launchConnect()
+
+        val connection = awaitConnection()
+        if (connection != null) return connection
+
+        morePlansExist = morePlansExist && routePlanner.hasMoreRoutes()
+      }
+
+      throw firstException!!
+    } finally {
+      for (plan in connectsInFlight) {
+        plan.cancel()
+      }
+    }
+  }
+
+  private fun launchConnect() {
+    if (!morePlansExist) return
+
+    val plan = try {
+      routePlanner.plan()
+    } catch (e: IOException) {
+      trackFailure(e)
+      return
+    }
+
+    connectsInFlight += plan
+
+    // Already connected? Enqueue the result immediately.
+    if (plan.isConnected) {
+      connectResults.put(ConnectResult(plan, null))
+      return
+    }
+
+    // Connect asynchronously.
+    val taskName = "$okHttpName connect ${routePlanner.address.url.redact()}"
+    taskRunner.newQueue().schedule(object : Task(taskName) {
+      override fun runOnce(): Long {
+        try {
+          plan.connect()
+          connectResults.put(ConnectResult(plan, null))
+        } catch (e: IOException) {
+          connectResults.put(ConnectResult(plan, e))
+        }
+        return -1L
+      }
+    })
+  }
+
+  private fun awaitConnection(): RealConnection? {
+    if (connectsInFlight.isEmpty()) return null
+
+    val completed = connectResults.poll(connectDelayMillis, TimeUnit.MILLISECONDS) ?: return null
+
+    connectsInFlight.remove(completed.plan)
+
+    val exception = completed.exception
+    if (exception != null) {
+      trackFailure(exception)
+      return null
+    }
+
+    return completed.plan.handleSuccess()
+  }
+
+  private fun trackFailure(exception: IOException) {
+    routePlanner.trackFailure(exception)
```

Comment:
Not needed at the moment, since it's only a counter, but should trackFailure be told the route also?

Reply:
https://datatracker.ietf.org/doc/html/rfc6555#section-4.2 Talks about stateful implementations

Reply:
Although changing mobile networks makes this complicated, would only apply until the network changes.

Reply:
I’ll add it when we need it!